corehq/apps/translations/integrations/transifex/project_migrator.py

```diff
@@ -0,0 +1,185 @@
+    def _source_app(self):
+        return get_app(self.domain, self.source_app_id)
+
+    def _update_context(self, translations):
+        """
+        update msgctxt for all POEntry objects replacing ids
+        :param translations: dict of lang code mapped to it list of POEntries
+        """
+        for po_entries in translations.values():
+            for po_entry in po_entries:
+                # make sure the format is as expected, if not skip
+                context_entries = po_entry.msgctxt.split(":")
+                if len(context_entries) == 3:
+                    resource_id = context_entries[-1]
+                    # replace if we have been asked to replace it
+                    if resource_id in self.id_mapping:
+                        po_entry.msgctxt = po_entry.msgctxt.replace(resource_id, self.id_mapping[resource_id])
+
+    def _upload_new_translations(self, translations):
+        responses = {}
+        # the project source lang, which is the app default language should be the first to update.
+        # HQ keeps the default lang on top and hence it should be the first one here
+        assert translations.keys()[0] == self.target_app_default_lang
+        for lang_code in translations:
+            responses[lang_code] = self._upload_translation(translations[lang_code], lang_code)
+        return responses
+
+    def _upload_translation(self, translations, lang_code):
+        po = polib.POFile()
+        po.check_for_duplicates = False
+        po.metadata = self.metadata
+        po.extend(translations)
+        with tempfile.NamedTemporaryFile() as temp_file:
+            po.save(temp_file.name)
+            temp_file.seek(0)
+            if lang_code == self.target_app_default_lang:
+                return self.client.upload_resource(temp_file.name, "Menus_and_forms", "Menus_and_forms",
+                                                   update_resource=True)
+            else:
+                return self.client.upload_translation(temp_file.name, "Menus_and_forms", "Menus_and_forms",
+                                                      lang_code)
+
+    @property
+    def metadata(self):
+        now = str(datetime.datetime.now())
+        return {
+            'App-Id': self.target_app_id,
+            'PO-Creation-Date': now,
+            'MIME-Version': '1.0',
+            'Content-Type': 'text/plain; charset=utf-8',
+            'Language': self.target_app_default_lang
+        }
+
+    @cached_property
+    def target_app_default_lang(self):
+        return self._target_app.default_language
+
+    @cached_property
+    def _target_app(self):
+        return get_app(self.domain, self.target_app_id)
+
+    @cached_property
+    def get_project_source_lang(self):
+        return self.client.project_details().json()['source_language_code']
+
+    @cached_property
+    def source_app_default_lang(self):
+        return self._source_app.default_language
+
+
+class ProjectMigrationValidator(object):
```

Comment:
It looks like this is only used from within the form, might make sense to move this logic there as part of the `clean` method.  As-is, validation happens in two different places, with different mechanisms for informing the user of the error.  If you make `_ensure_same_source_lang` to the form, you can use `ValidationError` which would be consistent with the other types of errors that are raised.

Reply:
the intention for doing this was to separate out logic for validating the migration from the requirements of the form.
For ex: if we were to run this as a background process somewhere where the details are provided more reliably and not via a user uploaded form, the validation logic for the migration itself would be available for us to run before running the real migration.